File: modules/version-consistency-checker/src/version_consistency_checker/providers/mise.py
import subprocess
from pathlib import Path

from version_consistency_checker.models.mise import MiseTools
from version_consistency_checker.providers.base import BaseProvider
from version_consistency_checker.utils import resolve_bin


class MiseProvider(BaseProvider):
  bin_path: Path
  project_dir: Path

  def __init__(self, project_dir: Path, bin_path: Path | None = None) -> None:
    self.project_dir = project_dir

    if bin_path is not None:
      self.bin_path = bin_path
    else:
      mise_path = resolve_bin.resolve_binary_path("mise")
      if mise_path is None:
        err_msg = "mise is not found in PATH"
        raise RuntimeError(err_msg)

      self.bin_path = mise_path

  def fetch_tool_versions(self) -> dict[str, str]:
    args: list[Path | str] = [self.bin_path, "list", "--json"]
    result = subprocess.run(
      args,
      capture_output=True,
      text=True,
      cwd=self.project_dir,
    )

    if result.returncode != 0:
      err_msg = f"failed to run mise: {result.stderr}"
      raise RuntimeError(err_msg)

    fetched: dict[str, str] = {}
    mise_tools = MiseTools.model_validate_json(result.stdout)
    for name, tools in mise_tools:
      for tool in tools:
        if not tool.active:
          continue

        fetched[name] = tool.version
        break

    return fetched

File: modules/version-consistency-checker/src/providers/mise.ts
import type { Result } from "@cffnpwr/result-ts";

import { Err, Ok } from "@cffnpwr/result-ts";
import { type } from "arktype";

import type { OptionalOption, Provider, ProviderFactory } from "./base.ts";

import { resolveBin } from "../utils/resolveBin.ts";

import { ProviderError } from "./errors.ts";

const miseTool = type({
  version: "string",
  "requested_version?": "string",
  install_path: "string",
  "source?": { type: "string", path: "string" },
  installed: "boolean",
  active: "boolean",
});

const parseMiseOutput = type("string.json.parse").pipe(
  type({ "[string]": miseTool.array() }),
);

const OPTION_PROJECT_DIR = {
  required: false,
  name: "--project-dir",
  type: "string",
  default: ".",
  help: "Path to the project directory containing mise.toml",
} as const satisfies OptionalOption;

const OPTION_MISE_BIN = {
  required: false,
  name: "--mise-bin",
  type: "string",
  help: "Path to the mise binary (default: resolved from PATH)",
} as const satisfies OptionalOption;

const createMiseProvider = (projectDir: string, miseBin: string): Provider => {
  return {
    async fetchToolVersions(): Promise<
      Result<Record<string, string>, ProviderError>
    > {
      const spawnResult = Bun.spawnSync([miseBin, "list", "--json"], {
        cwd: projectDir,
      });

      if (spawnResult.exitCode !== 0) {
        return Err(
          new ProviderError(
            `failed to run mise: ${spawnResult.stderr.toString()}`,
          ),
        );
      }

      const validated = parseMiseOutput(spawnResult.stdout.toString());
      if (validated instanceof type.errors) {
        return Err(
          new ProviderError(
            `failed to parse mise output: ${validated.summary}`,
          ),
        );
      }

      const versions: Record<string, string> = {};
      for (const [name, tools] of Object.entries(validated)) {
        const active = tools.find((t) => t.active);
        if (active !== undefined) {
          versions[name] = active.version;
        }
      }

      return Ok(versions);
    },
  };
};

export const miseProviderFactory: ProviderFactory = {
  sourceName: "mise",

  getOptions(): OptionalOption[] {
    return [OPTION_PROJECT_DIR, OPTION_MISE_BIN];
  },

  create(args: Record<string, unknown>): Result<Provider, ProviderError> {
    const projectDirArg = args[OPTION_PROJECT_DIR.name.slice(2)];
    const projectDir = typeof projectDirArg === "string" && projectDirArg !== ""
      ? projectDirArg
      : OPTION_PROJECT_DIR.default;

    const miseBinArg = args[OPTION_MISE_BIN.name.slice(2)];
    if (typeof miseBinArg === "string" && miseBinArg !== "") {
      return Ok(createMiseProvider(projectDir, miseBinArg));
    }

    const resolved = resolveBin("mise");
    if (resolved.isNone()) {
      return Err(new ProviderError("mise is not found in PATH"));
    }

    return Ok(createMiseProvider(projectDir, resolved.unwrap()));
  },
};
